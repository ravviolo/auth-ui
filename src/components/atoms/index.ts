export { Button } from './Button/Button';
export { Checkbox } from './Checkbox/Checkbox';
export { Divider } from './Divider/Divider';
export { Heading } from './Heading/Heading';
export { IconButton } from './IconButton/IconButton';
export { InputField } from './InputField/InputField';
export { InputLabel } from './InputLabel/InputLabel';
export { TextButton } from './TextButton/TextButton';
