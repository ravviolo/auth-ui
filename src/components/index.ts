export * from './templates/LoginTemplate/LoginTemplate';
export * from './templates/SignUpTemplate/SignUpTemplate';
