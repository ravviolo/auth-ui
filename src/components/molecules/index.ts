export { AuthFooter } from './AuthFooter/AuthFooter';
export { AuthHeader } from './AuthHeader/AuthHeader';
export { AuthSocials } from './AuthSocials/AuthSocials';
export { EmailInput } from './EmailInput/EmailInput';
export { PasswordInput } from './PasswordInput/PasswordInput';
