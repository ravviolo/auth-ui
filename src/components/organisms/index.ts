export * from './LoginForm/LoginForm';
export * from './SignUpForm/SignUpForm';
