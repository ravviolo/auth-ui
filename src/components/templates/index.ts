export * from './LoginTemplate/LoginTemplate';
export * from './SignUpTemplate/SignUpTemplate';
