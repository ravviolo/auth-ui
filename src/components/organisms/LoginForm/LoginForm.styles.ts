import styled from 'styled-components';

import { Flex, TextButton } from 'components/atoms';
import { colors } from 'theme';

export const Container = styled(Flex)`
  gap: 1em;
`;

export const PasswordControl = styled(Flex)`
  gap: 0.8em;
`;

export const CheckboxContainer = styled(Flex)`
  gap: 0.8em;
  align-items: center;

  & > input {
    height: 1rem;
    width: 1rem;
  }
`;

export const ResetPasswordButton = styled(TextButton)`
  color: ${colors.textLight};
  margin-left: auto;
  transform: translateY(-40%);
`;
