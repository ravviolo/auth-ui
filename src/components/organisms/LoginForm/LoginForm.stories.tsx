import { ComponentMeta, ComponentStory } from '@storybook/react';

import { LoginForm } from 'components/organisms';

export default {
  title: 'Organisms/LoginForm',
  component: LoginForm,
} as ComponentMeta<typeof LoginForm>;

export const Default: ComponentStory<typeof LoginForm> = (args) => (
  <LoginForm {...args} />
);

Default.storyName = 'LoginForm';
