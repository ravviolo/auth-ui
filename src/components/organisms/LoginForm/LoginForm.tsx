import { useRef } from 'react';

import { Button, Checkbox, InputLabel } from 'components/atoms';
import { EmailInput, PasswordInput } from 'components/molecules';

import * as S from './LoginForm.styles';

interface Props {
  onSubmitLoginForm: (
    email: string,
    password: string,
    isPasswordRemembered: boolean
  ) => void;
  onClickResetPassword: () => void;
}

export const LoginForm = ({
  onSubmitLoginForm,
  onClickResetPassword,
}: Props) => {
  const emailRef = useRef<HTMLInputElement>(null);
  const passwordRef = useRef<HTMLInputElement>(null);
  const rememberPasswordRef = useRef<HTMLInputElement>(null);

  const handleLogin: React.FormEventHandler<HTMLFormElement> = (e) => {
    e.preventDefault();

    const email = emailRef.current?.value;
    const password = passwordRef.current?.value;
    const isPasswordRemembered = !!rememberPasswordRef.current?.checked;

    if (email && password) {
      onSubmitLoginForm(email, password, isPasswordRemembered);
    }
  };

  return (
    <form onSubmit={handleLogin}>
      <S.Container direction="column">
        <EmailInput ref={emailRef} id="email" name="email" />
        <S.PasswordControl direction="column">
          <PasswordInput ref={passwordRef} id="password" name="password" />
          <S.CheckboxContainer direction="row">
            <Checkbox
              ref={rememberPasswordRef}
              id="rememberPassword"
              name="rememberPassword"
            />
            <InputLabel htmlFor="rememberPassword" label="Remember me?" />
          </S.CheckboxContainer>
        </S.PasswordControl>
        <Button label="Login" testId="login-btn" type="submit" />

        <S.ResetPasswordButton
          label="Forgot Password?"
          type="button"
          onClick={onClickResetPassword}
        />
      </S.Container>
    </form>
  );
};
