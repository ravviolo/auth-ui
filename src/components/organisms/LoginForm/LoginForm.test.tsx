import { render, screen, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { create } from 'react-test-renderer';

import { LoginForm } from 'components/organisms';

import { Props as LoginFormProps } from './LoginForm';

describe('LoginForm', () => {
  const mockOnSubmitLoginForm = jest.fn<
    LoginFormProps['onSubmitLoginForm'],
    [string, string, boolean]
  >();
  const mockOnClickResetPassword = jest.fn<
    LoginFormProps['onClickResetPassword'],
    []
  >();
  const testEmail = 'test@test.com';
  const testPassword = 'testpassword';

  const props: LoginFormProps = {
    testId: 'test-login-form',
    onClickResetPassword: mockOnClickResetPassword,
    onSubmitLoginForm: mockOnSubmitLoginForm,
  };

  it('renders email and password inputs', () => {
    render(<LoginForm {...props} />);

    const emailInput = screen.getByTestId('email-input-login-form');
    const passwordInput = screen.getByTestId('password-input-login-form');

    expect(emailInput).toBeInTheDocument();
    expect(passwordInput).toBeInTheDocument();
  });

  it("renders checkbox with 'Remember me?' label", () => {
    render(<LoginForm {...props} />);

    const checkbox = screen.getByTestId('remember-password-checkbox');
    const checkboxLabel = screen.getByTestId(
      'input-label-remember-password-checkbox'
    );

    expect(checkboxLabel).toHaveTextContent(/Remember me?/);
    expect(checkboxLabel).toHaveAttribute('for', checkbox.getAttribute('id'));
  });

  it('renders initially unchecked checkbox', () => {
    render(<LoginForm {...props} />);

    const checkbox = screen.getByTestId('remember-password-checkbox');

    expect(checkbox).not.toBeChecked();
  });

  it("renders submit button with 'Login' text", () => {
    render(<LoginForm {...props} />);

    const submitButton = screen.getByTestId('btn-submit-login');

    expect(submitButton).toHaveAttribute('type', 'submit');
    expect(submitButton).toHaveTextContent('Login');
  });

  it('submits form with non-empty input fields and checkbox unchecked - runs submit handler once', async () => {
    render(<LoginForm {...props} />);

    const submitButton = screen.getByTestId('btn-submit-login');
    const emailInputField = within(
      screen.getByTestId('email-input-login-form')
    ).getByTestId('input-field-email');
    const passwordInputField = within(
      screen.getByTestId('password-input-login-form')
    ).getByTestId('input-field-password');

    await userEvent.type(emailInputField, testEmail);
    await userEvent.type(passwordInputField, testPassword);
    await userEvent.click(submitButton);

    expect(mockOnSubmitLoginForm).toHaveBeenCalledTimes(1);
    expect(mockOnSubmitLoginForm).toHaveBeenCalledWith(
      testEmail,
      testPassword,
      false
    );
  });

  it('submits form with non-empty input fields and checkbox checked - runs submit handler once', async () => {
    render(<LoginForm {...props} />);

    const submitButton = screen.getByTestId('btn-submit-login');
    const emailInputField = within(
      screen.getByTestId('email-input-login-form')
    ).getByTestId('input-field-email');
    const passwordInputField = within(
      screen.getByTestId('password-input-login-form')
    ).getByTestId('input-field-password');
    const checkbox = screen.getByTestId('remember-password-checkbox');

    await userEvent.type(emailInputField, testEmail);
    await userEvent.type(passwordInputField, testPassword);
    await userEvent.click(checkbox);
    await userEvent.click(submitButton);

    expect(mockOnSubmitLoginForm).toHaveBeenCalledTimes(1);
    expect(mockOnSubmitLoginForm).toHaveBeenCalledWith(
      testEmail,
      testPassword,
      true
    );
  });

  it("doesn't submit form with one or more empty input fields", async () => {
    render(<LoginForm {...props} />);

    const submitButton = screen.getByTestId('btn-submit-login');

    await userEvent.click(submitButton);

    expect(mockOnSubmitLoginForm).not.toHaveBeenCalled();

    const emailInputField = within(
      screen.getByTestId('email-input-login-form')
    ).getByTestId('input-field-email');

    await userEvent.type(emailInputField, testEmail);
    await userEvent.click(submitButton);

    expect(mockOnSubmitLoginForm).not.toHaveBeenCalled();
  });

  it("renders reset password button with 'Forgot Password?' text", () => {
    render(<LoginForm {...props} />);

    const resetPasswordButton = screen.getByTestId('btn-reset-password');

    expect(resetPasswordButton).toHaveAttribute('type', 'button');
    expect(resetPasswordButton).toHaveTextContent('Forgot Password?');
  });

  it('reset password button handles click events - runs click handler once', async () => {
    render(<LoginForm {...props} />);

    const resetPasswordButton = screen.getByTestId('btn-reset-password');

    await userEvent.click(resetPasswordButton);

    expect(mockOnClickResetPassword).toHaveBeenCalledTimes(1);
  });

  it('matches snapshot', () => {
    const tree = create(<LoginForm {...props} />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
