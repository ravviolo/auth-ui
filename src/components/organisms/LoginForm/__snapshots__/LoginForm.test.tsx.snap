// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm matches snapshot 1`] = `
<form
  data-testid="test-login-form"
  onSubmit={[Function]}
>
  <div
    className="sc-jSMfEi bQrDqy sc-hAZoDl bWsHpL"
    data-testid="flex-login-form"
    direction="column"
  >
    <div
      className="sc-jSMfEi bQrDqy sc-breuTD jjaaRC"
      data-testid="email-input-login-form"
      direction="column"
    >
      <label
        className="sc-papXJ daxRFt"
        data-testid="input-label-email"
        htmlFor="email"
      >
        Email
      </label>
      <input
        className="sc-ftvSup dLJLzu"
        data-testid="input-field-email"
        id="email"
        name="email"
        type="email"
      />
    </div>
    <div
      className="sc-jSMfEi bQrDqy sc-fnykZs hELmKH"
      data-testid="flex-login-form-password-control"
      direction="column"
    >
      <div
        className="sc-jSMfEi bQrDqy sc-ksZaOG cRa-dNN"
        data-testid="password-input-login-form"
        direction="column"
      >
        <label
          className="sc-papXJ daxRFt"
          data-testid="input-label-password"
          htmlFor="password"
        >
          Password
        </label>
        <input
          className="sc-ftvSup dLJLzu"
          data-testid="input-field-password"
          id="password"
          name="password"
          type="password"
        />
      </div>
      <div
        className="sc-jSMfEi iChCMC sc-fEOsli dPSjkT"
        data-testid="flex-login-form-checkbox"
        direction="row"
      >
        <input
          className="sc-gsnTZi kVuJZl"
          data-testid="remember-password-checkbox"
          id="rememberPassword"
          name="rememberPassword"
          type="checkbox"
        />
        <label
          className="sc-papXJ daxRFt"
          data-testid="input-label-remember-password-checkbox"
          htmlFor="rememberPassword"
        >
          Remember me?
        </label>
      </div>
    </div>
    <button
      className="sc-bczRLJ jXOWFG"
      data-testid="btn-submit-login"
      type="submit"
    >
      Login
    </button>
    <button
      className="sc-jqUVSM ebpvV sc-bjUoiL hqHESO"
      data-testid="btn-reset-password"
      onClick={[MockFunction]}
      type="button"
    >
      Forgot Password?
    </button>
  </div>
</form>
`;
